### ISSUE TITLE
Parse SQL Case Expressions properly enabling better laballing of charts, error messages, match weights to apply to different cols etc

### ISSUE THREAD

**Issue by RobinL** [moj-analytical-services/splink#240]
`sqlglot` provides a MIT-licenced, dependency free way of parsing multiple SQL dialects.  I've experimented and I think it could be very useful:

```python
import sqlglot
from sqlglot.expressions import Case, Column


def get_columns_used_from_expr(expr):
    column_names = []
    for tup in expr.walk():
        item, parent, key = tup
        if type(item) is Column:
            column_names.append(item.sql())
    return column_names


def parse_else_value(item):
    res = {}
    case_args = item.args
    else_stmt = case_args["default"]
    else_stmt_value = else_stmt.this

    res["literal_value"] = else_stmt_value
    res["sql"] = f"ELSE {else_stmt_value} END"
    res["cols_used"] = []
    return res


def parse_if_value(item):
    res = {}
    case_args = item.args
    if_stmt_value = case_args["true"].this

    res["literal_value"] = if_stmt_value
    sql = item.sql()

    import re

    sql = re.sub(r"^CASE ", "", sql)

    sql = re.sub(r" END$", "", sql)

    res["sql"] = sql
    cols_used = get_columns_used_from_expr(item)
    res["cols_used"] = cols_used
    return res


def parse_case_statement(sql_expression):

    # Python 3.7 and CPython 3.6 dicts guaranteed to iterate in order of insertion
    case_statement_definition = {}

    parsed = sqlglot.parse_one(sql_expression)

    for tup in parsed.walk():

        item, parent, key = tup

        if type(item) is Case:
            all_cols_used = get_columns_used_from_expr(item)

            case_args = item.args

            ifs = case_args["ifs"]
            for this_if in ifs:

                if_dict = parse_if_value(this_if)
                val = if_dict["literal_value"]
                case_statement_definition[val] = if_dict

            else_dict = parse_else_value(item)
            val = else_dict["literal_value"]
            case_statement_definition[val] = else_dict

    return case_statement_definition


smt = (
    "case when tab.first = 'first' then 1 when second = 200 then 2 else 3 end as mycol"
)

result = parse_case_statement(smt)

import json

print(json.dumps(result, indent=4))
```

Results in 

```python

    "1": {
        "literal_value": "1",
        "sql": "WHEN tab.first = 'first' THEN 1",
        "cols_used": [
            "tab.first"
        ]
    },
    "2": {
        "literal_value": "2",
        "sql": "WHEN second = 200 THEN 2",
        "cols_used": [
            "second"
        ]
    },
    "3": {
        "literal_value": "3",
        "sql": "ELSE 3 END",
        "cols_used": []
    }
}
```

**Comment by RobinL** [comment]
This could be used to:

**Comment by RobinL** [comment]
Another useful snippet to show how the parsed expression tree works:
```
for tup in parsed.walk():

    # queue = [item, parent, key]
    item = tup[0]
    parent = tup[1]
    key = tup[2]

    try:
        item_sql = item.sql()

        print(f"{'    '*item.depth}{item_sql = }")
        print(f"{'    '*item.depth}{item.depth = }")
        print(f"{'    '*item.depth}{item.__class__ = }")

    except AttributeError:
        item_sql = None

    try:
        parent_sql = parent.sql()
    except AttributeError:
        parent_sql = None
```

**Comment by RobinL** [comment]
This could also probably be used to parse blocking rules, and write a function like:
```
num_comparisons_generated_by_blocking_rules()
```

**Comment by RobinL** [comment]
Closed by #261
